test/Microsoft.ML.TestFramework/BaseTestBaseline.cs

```diff
@@ -111,22 +111,61 @@ protected override void Initialize()
+        {
+            // Sometimes different configuration will have combination.
+            // for example: one test can run on windows, have different result both
+            // on x64 vs x86 and dotnet core 3.1 vs others, so we have 4 combination:
+            // x64-netcore3.1, x86-netcore3.1, x64-rest, x86-rest. In some cases x64 vs x86
+            // have different results, in some cases netcore 3.1 vs rest have different results,
+            // the most complicate situation is 12 combinations (x64 vs x86, netcoreapp3.1 vs rest,
+            // win vs linux vs osx) have different results.
+            // So use list of string to return different configurations and test will try to search
+            // through this list and use the one file first found, make sure we don't have baseline file
+            // at different configuration folders.
+            var configurationDirs = new List<string>();
+
+            // Use netcore 3.1 result file if necessary.
+            // The small difference comes from CPUMath using different instruction set:
+            // 1. net framework and net core 2.1 uses CpuMathUtils.netstandard that uses SSE instruction set;
+            // 2. net core 3.1 uses CpuMathUtils.netcoreapp that uses AVX, SSE or direct floating point calculation
+            // depending on hardward avaibility.
+            // AVX and SSE generates slightly different result due to nature of floating point math.
+            // So Ideally we should adding AVX support at CPUMath native library, 
+            // use below issue to track: https://github.com/dotnet/machinelearning/issues/5044
+            // don't need netcoreapp21 as this is the default case
+            if (AppDomain.CurrentDomain.GetData("FX_PRODUCT_VERSION") != null)
+                configurationDirs.Add("netcoreapp31");
+
+            // don't need x64 as this is the default case
+            if (!Environment.Is64BitProcess)
+                configurationDirs.Add("x86");
+
             if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
-                return "osx-x64";
+                configurationDirs.Add("osx-x64");
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
-                return "linux-x64";
+                configurationDirs.Add("linux-x64");
             if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                 if (Environment.Is64BitProcess)
-                    return "win-x64";
+                {
+                    configurationDirs.Add("win-x64");
+
+                    if (AppDomain.CurrentDomain.GetData("FX_PRODUCT_VERSION") != null)
+                        configurationDirs.Add($"win-x64-netcoreapp31");
+                    else
+                        configurationDirs.Add($"win-x64-netcoreapp21");
+                }
                 else
-                    return "win-x86";
+                {
+                    configurationDirs.Add("win-x86");
 
```

Comment:
The order of these should be reversed. 
We should check from the most specific to the least specific. The structure in the folder names we have is <os>-<processor>-<framework>. So we should check win-x86-netcoreapp31 first, followed by win-x86 followed by 
 common.

Reply:
yes, that make sense, thanks

---
In reply to: [413234549](https://github.com/dotnet/machinelearning/pull/5047#discussion_r413234549) [](ancestors = 413234549)